JS: Add 'prototype pollution in utility function' query

asgerf opened the pull request:
Adds a query that looks for recursive merge functions that are susceptible to prototype pollution. It doesn't check for user-controlled inputs.

For example, it flags this function as vulnerable:
```js
function merge(dst, src) {
    for (let key in src) {
        if (!src.hasOwnProperty(key)) continue;
        if (isObject(dst[key])) {
            merge(dst[key], src[key]);
        } else {
            dst[key] = src[key];
        }
    }
}
```
It looks for a property enumeration (the `for-in` loop above) and a dynamic property write (`dst[key] = src[key]` above), where three flow paths can be found:
- From `key` to the base of the dynamic property write
- From `key` to the property name of the dynamic property write
- From `src[key]` to the right-hand side fo the dynamic property write

It's relatively expensive as there are a lot of sources and sinks, and the graph pruning does not take into account that we require all three paths to exist. For example, if the base of a dynamic property write is definitely unreachable from a source, it might still try to find paths to the property name and RHS of that property write, even though these paths will never result in an alert.

The [latest evaluation](https://git.semmle.com/asger/dist-compare-reports/tree/prototype-pollution-utility_1578590273313) shows the first serious conflict I've seen between wall-clock time and tuple counts/DPMs, and I'd like to run a few more evaluations to determine which one tells the truth.

erik-krogh commented:
What about the FP rate? 

You get quite a few results just on `nightly.slugs`, so I worry that the results will be noisy once deployed on LGTM.

asgerf commented:
There's a FP in jQuery due to our imprecise treatment of overloading in general. The rest are TPs as far as I can tell (although whether any user-controlled inputs flow there remains TBD).

asgerf commented:
[Re-running the evaluation](https://git.semmle.com/asger/dist-compare-reports/tree/prototype-pollution-utility_1578758392977) shows the wall clock timing was a fluke. +1 for tuple counts/DPM.

esbena commented on javascript/ql/src/Security/CWE-400/PrototypePollutionUtility.ql:
````diff
@@ -0,0 +1,560 @@
+ * ```
+ *
+ * This configuration is used to find three separate data flow paths originating
+ * from a property enumeration, all leading to the same dynamic property write.
````
> all leading to the same dynamic property write.

That is not enforced by this configuration, but by `isPrototypePollutingAssignment` below, right?

This is a quite special use of a configuration, can we add a comment about how this configuration is not a standalone configuration like most other configurations?

asgerf replied:
Added a comment at the bottom of this qldoc

esbena commented on javascript/ql/src/Security/CWE-400/PrototypePollutionUtility.ql:
```diff
@@ -0,0 +1,560 @@
+  override predicate isSink(DataFlow::Node node, FlowLabel label) {
+    label instanceof UnsafePropLabel and
+    (
+      dynamicPropWrite(node, _, _) or
```
Can you explain why the last conjunct of the `where` below is not inlined here instead?
```
  (
    getANodeLeadingToBaseBase(base) instanceof ObjectLiteralNode
    or
    not getANodeLeadingToBaseBase(base) instanceof ObjectCreateNullCall
  )
```

asgerf replied:
Because `getANodeLeadingToBaseBase` depends on `hasFlow`.

Instead of forward-tracking from `Object.create(null)` we track backwards from potential sinks as a post-processing step.

esbena commented on javascript/ql/src/Security/CWE-400/PrototypePollutionUtility.ql:
```diff
@@ -0,0 +1,560 @@
+ * `x -> x[p]` and `p -> x[p]`.
+ *
+ * Note that in the above example, the flow from `key` to the base of the write (`dst`)
+ * requires stepping through the recursive call.
```
Can you add a comment to the crucial part of this query that demands this recursive call?

asgerf replied:
Expanded on the comment a little bit, to explain why shallow copying isn't interesting.

I'm not sure I understand your question though. There isn't a part of the query that demands a recursive call explicitly, but it's known that shallow copying is safe, and recursive copying in unsafe, so it's worth pointing out how the query distinguishes those two. The comment aims to explain how the distinction is made: shallow copies lack the path `key -> dst`.

esbena commented on javascript/ql/src/Security/CWE-400/PrototypePollutionUtility.ql:
```diff
@@ -0,0 +1,560 @@
+                accidental modification of a built-in prototype object.
+ * @kind path-problem
+ * @problem.severity warning
+ * @precision high
```
:tada: Very nice!

asgerf commented on javascript/ql/src/Security/CWE-400/PrototypePollutionUtility.ql:
```diff
@@ -0,0 +1,560 @@
+      result = AccessPath::getAReferenceTo(root, path)
+    )
+    or
+    result = getSourceObject().getALocalSource()
```
Opened an issue for it

asgerf commented on javascript/ql/src/semmle/javascript/GlobalAccessPaths.qll:
```diff
@@ -115,6 +145,9 @@ module AccessPath {
       result = join(fromReference(prop.getBase(), root), prop.getPropertyName())
+      or
+      isLikelyDynamicArrayAccess(prop) and
+      result = join(fromReference(prop.getBase(), root), "[number]")
```
Added a doc comment on `AccessPath` module and mentioned the `[number]` property there.

asgerf commented:
Added a change note, and rebased to fix change note conflict.

I also changed the CWE numbers. Actually I'd like to move both the prototype pollution queries to CWE-471 (modification of assumed-immutable data) but I think I'll do that in another PR.

esbena commented:
Excellent. 
Ping @mchammer01 for a doc review

Two meta tasks:

- have you pinged maintainers or our security team about the (relevant) newly discovered vulnerabilities?
- will you update the tracking issues for the 10 CVEs that inspired this query?

asgerf commented on javascript/ql/src/Security/CWE-400/PrototypePollutionUtility.qhelp:
It's technical jargon - an "own property" means something specific in JavaScript.

asgerf commented on javascript/ql/src/Security/CWE-400/PrototypePollutionUtility.qhelp:
It's technical jargon - an "own property" means something specific in JavaScript.

asgerf commented on javascript/ql/src/Security/CWE-400/PrototypePollutionUtility.qhelp:
Hm, again it's fairly technical and I think it's easier to understand the other way.
